ui/components/src/ThresholdsEditor/ThresholdsEditor.tsx

```diff
@@ -0,0 +1,238 @@
+        if (draft.steps) {
+          draft.steps.splice(i, 1);
+        }
+      });
+      onChange(updatedThresholds);
+    }
+  };
+
+  const addThresholdInput = (): void => {
+    focusRef.current = true;
+    if (thresholds === undefined) {
+      onChange({
+        steps: [{ value: DEFAULT_STEP }],
+      });
+    } else {
+      onChange(
+        produce(thresholds, (draft) => {
+          const steps = draft.steps;
+          if (steps?.length) {
+            const lastStep = steps[steps.length - 1];
+            const color = palette[steps.length] ?? getRandomColor(); // we will assign color from the palette first, then generate random color
+            steps.push({ color, value: (lastStep?.value ?? 0) + DEFAULT_STEP }); // set new threshold value to last step value + 10
+          } else if (steps) {
+            steps.push({ value: DEFAULT_STEP });
+          }
+        })
+      );
+    }
+  };
+
+  const handleModeChange = (event: React.MouseEvent, value: string): void => {
+    if (thresholds !== undefined) {
+      onChange(
+        produce(thresholds, (draft) => {
+          draft.mode = value === 'Percent' ? 'Percent' : undefined;
+        })
+      );
+    }
+  };
+
+  return (
+    <OptionsEditorGroup
+      title="Thresholds"
+      icon={
+        <IconButton size="small" aria-label="add threshold" onClick={addThresholdInput}>
+          <PlusIcon />
+        </IconButton>
+      }
+    >
+      <ToggleButtonGroup
+        exclusive
+        value={thresholds?.mode ?? 'absolute'}
+        onChange={handleModeChange}
+        sx={{ height: '36px', marginLeft: 'auto' }}
+      >
+        <ToggleButton aria-label="absolute" value="Absolute">
+          <InfoTooltip description="Absolute">#</InfoTooltip>
+        </ToggleButton>
+        <ToggleButton aria-label="percent" value="Percent">
+          <InfoTooltip description="Percentage means thresholds relative to min & max">%</InfoTooltip>
+        </ToggleButton>
+      </ToggleButtonGroup>
+      {steps &&
+        steps
+          .map((step, i) => (
+            <ThresholdInput
+              inputRef={i === steps.length - 1 ? recentlyAddedInputRef : undefined}
+              key={i}
+              label={`T${i + 1}`}
+              color={step.color ?? palette[i] ?? defaultThresholdColor}
+              value={step.value}
+              onColorChange={(color) => handleThresholdColorChange(color, i)}
+              onChange={(e) => {
+                handleThresholdValueChange(e, i);
+              }}
+              onDelete={() => {
+                deleteThreshold(i);
+              }}
+              onBlur={handleThresholdBlur}
+            />
+          ))
+          .reverse()}
+      <Stack flex={1} direction="row" alignItems="center" spacing={1}>
+        <ThresholdColorPicker label="default" color={defaultThresholdColor} onColorChange={handleDefaultColorChange} />
+        <Typography>Default</Typography>
+      </Stack>
+    </OptionsEditorGroup>
+  );
+}
+
```

Comment:
This is already a relatively big file. Is it work pulling the `ThresholdInput` component out into a separate file?